File: fui_system/src/platform/qt/qt_wrapper/qfiledialog.rs
use crate::platform::qt::qt_wrapper::{QString, QStringList};

pub struct QFileDialog;

impl QFileDialog {
    pub fn get_open_file_name(
        caption: QString,
        dir: QString,
        filter: QString,
        selected_filter: &mut QString,
        options: i32,
    ) -> Option<QString> {
        unsafe {
            let qstring_ptr = crate::platform::qt::qt_wrapper::QFileDialog_getOpenFileName(
                std::ptr::null_mut(),
                caption.this,
                dir.this,
                filter.this,
                selected_filter.this,
                options,
            );
            if qstring_ptr.is_null() {
                None
            } else {
                Some(QString {
                    this: qstring_ptr,
                    is_owned: true,
                })
            }
        }
    }

    pub fn get_open_file_names(
        caption: QString,
        dir: QString,
        filter: QString,
        selected_filter: &mut QString,
        options: i32,
    ) -> QStringList {
        unsafe {
            let qstringlist_ptr = crate::platform::qt::qt_wrapper::QFileDialog_getOpenFileNames(
                std::ptr::null_mut(),
                caption.this,
                dir.this,
                filter.this,
                selected_filter.this,
                options,
            );
            QStringList {
                this: qstringlist_ptr,
            }
        }
    }

    pub fn get_existing_directory(caption: QString, dir: QString, options: i32) -> Option<QString> {
        unsafe {
            let qstring_ptr = crate::platform::qt::qt_wrapper::QFileDialog_getExistingDirectory(
                std::ptr::null_mut(),
                caption.this,
                dir.this,
                options,
            );
            if qstring_ptr.is_null() {
                None
            } else {
                Some(QString {
                    this: qstring_ptr,
                    is_owned: true,
                })
            }
        }
    }

    pub fn get_save_file_name(
        caption: QString,
        dir: QString,
        filter: QString,
        selected_filter: &mut QString,
        options: i32,
    ) -> Option<QString> {
        unsafe {
            let qstring_ptr = crate::platform::qt::qt_wrapper::QFileDialog_getSaveFileName(
                std::ptr::null_mut(),
                caption.this,
                dir.this,
                filter.this,
                selected_filter.this,
                options,
            );
            if qstring_ptr.is_null() {
                None
            } else {
                Some(QString {
                    this: qstring_ptr,
                    is_owned: true,
                })
            }
        }
    }
}

File: fui_system/src/platform/qt/qt_wrapper/cpp/qfiledialog.cpp
#include "qfiledialog.h"
#include <QWidget>
#include <QString>
#include <QFileDialog>

void *QFileDialog_getOpenFileName(void *parent,
                              void *caption, void *dir,
                              void *filter, void *selected_filter,
                              int options)
{
    QWidget *qparent = static_cast<QWidget *>(parent);
    const QString *qcaption = static_cast<const QString *>(caption);
    const QString *qdir = static_cast<const QString *>(dir);
    const QString *qfilter = static_cast<const QString *>(filter);
    QString *qselectedfilter = static_cast<QString *>(selected_filter);
    const QFileDialog::Options qoptions = static_cast<const QFileDialog::Options>(options);
    QString result = QFileDialog::getOpenFileName(qparent,
                                        *qcaption, *qdir,
                                        *qfilter, qselectedfilter,
                                        qoptions);
    if (result.isNull()) {
        return nullptr;
    }
    return static_cast<void *>(new QString(std::move(result)));
}

void *QFileDialog_getOpenFileNames(void *parent,
                                   void *caption, void *dir,
                                   void *filter, void *selected_filter,
                                   int options)
{
    QWidget *qparent = static_cast<QWidget *>(parent);
    const QString *qcaption = static_cast<const QString *>(caption);
    const QString *qdir = static_cast<const QString *>(dir);
    const QString *qfilter = static_cast<const QString *>(filter);
    QString *qselectedfilter = static_cast<QString *>(selected_filter);
    const QFileDialog::Options qoptions = static_cast<const QFileDialog::Options>(options);
    QStringList result = QFileDialog::getOpenFileNames(qparent,
                                                   *qcaption, *qdir,
                                                   *qfilter, qselectedfilter,
                                                   qoptions);
    return static_cast<void *>(new QStringList(std::move(result)));
}

void *QFileDialog_getExistingDirectory(void *parent,
                                       void *caption, void *dir,
                                       int options)
{
    QWidget *qparent = static_cast<QWidget *>(parent);
    const QString *qcaption = static_cast<const QString *>(caption);
    const QString *qdir = static_cast<const QString *>(dir);
    const QFileDialog::Options qoptions = static_cast<const QFileDialog::Options>(options);
    QString result = QFileDialog::getExistingDirectory(qparent,
                                                       *qcaption, *qdir,
                                                       qoptions);
    if (result.isNull()) {
        return nullptr;
    }
    return static_cast<void *>(new QString(std::move(result)));
}

void *QFileDialog_getSaveFileName(void *parent,
                                  void *caption, void *dir,
                                  void *filter, void *selected_filter,
                                  int options)
{
    QWidget *qparent = static_cast<QWidget *>(parent);
    const QString *qcaption = static_cast<const QString *>(caption);
    const QString *qdir = static_cast<const QString *>(dir);
    const QString *qfilter = static_cast<const QString *>(filter);
    QString *qselectedfilter = static_cast<QString *>(selected_filter);
    const QFileDialog::Options qoptions = static_cast<const QFileDialog::Options>(options);
    QString result = QFileDialog::getSaveFileName(qparent,
                                                  *qcaption, *qdir,
                                                  *qfilter, qselectedfilter,
                                                  qoptions);
    if (result.isNull()) {
        return nullptr;
    }
    return static_cast<void *>(new QString(std::move(result)));
}
